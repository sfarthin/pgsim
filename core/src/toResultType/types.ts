import parse, { PrimitiveType } from "../parse";

export type Field = {
  name: string | null;
  type: PrimitiveType;
  isNullable: boolean;
};

export type Source = { name: string; fields: Field[] };
