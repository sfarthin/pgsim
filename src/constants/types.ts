// https://www.postgresql.org/docs/9.5/datatype.html
export const types = [
  {
    description: "signed eight-byte integer",
    name: "int8",
    alias: ["bigint"],
    maxParams: 0,
  },
  {
    description: "autoincrementing eight-byte integer",
    name: "bigserial",
    alias: [],
    maxParams: 0,
  },
  {
    description: "autoincrementing eight-byte integer",
    name: "serial8",
    alias: [],
    maxParams: 0,
  },
  {
    description: "variable-length bit string",
    name: "varbit",
    alias: ["bit varying"],
    maxParams: 1,
  },
  {
    description: "fixed-length bit string",
    name: "bit",
    alias: [],
    maxParams: 1,
  },
  {
    description: "logical Boolean (true/false)",
    name: "bool",
    alias: ["boolean"],
    maxParams: 0,
  },
  {
    description: "rectangular box on a plane",
    name: "box",
    alias: [],
    maxParams: 0,
  },
  {
    description: 'binary data ("byte array")',
    name: "bytea",
    alias: [],
    maxParams: 0,
  },

  {
    description: "variable-length character string",
    name: "varchar",
    alias: ["character varying"],
    maxParams: 1,
  },
  {
    description: "fixed-length character string",
    name: "bpchar",
    alias: ["character", "char"],
    maxParams: 1,
  },
  {
    description: "IPv4 or IPv6 network address",
    name: "cidr",
    alias: [],
    maxParams: 0,
  },
  {
    description: "circle on a plane",
    name: "circle",
    alias: [],
    maxParams: 0,
  },
  {
    description: "calendar date (year, month, day)",
    name: "date",
    alias: [],
    maxParams: 0,
  },
  {
    description: "double precision floating-point number (8 bytes)",
    name: "float8",
    alias: ["double precision", "float"],
    maxParams: 0,
  },
  {
    description: "IPv4 or IPv6 host address",
    name: "inet",
    alias: [],
    maxParams: 0,
  },
  {
    description: "signed four-byte integer",
    name: "int4",
    alias: ["integer", "int4", "int"],
    maxParams: 0,
  },
  {
    description: "time span",
    name: "interval",
    // This is kind of a hack, if multiple spaces or comments are put between these keywords,
    // our parser won't parse them correctly. In the future, we can probably improve this.
    alias: [
      "interval year",
      "interval month",
      "interval day",
      "interval hour",
      "interval minute",
      "interval second",
      "interval year to month",
      "interval day to hour",
      "interval day to month",
      "interval day to minute",
      "interval day to second",
      "interval hour to minute",
      "interval hour to second",
      "interval minute to second",
    ],
    maxParams: 0,
  },
  {
    description: "textual JSON data",
    name: "json",
    alias: [],
    maxParams: 0,
  },
  {
    description: "binary JSON data, decomposed",
    name: "jsonb",
    alias: [],
    maxParams: 0,
  },
  {
    description: "infinite line on a plane",
    name: "line",
    alias: [],
    maxParams: 0,
  },
  {
    description: "line segment on a plane",
    name: "lseg",
    alias: [],
    maxParams: 0,
  },

  {
    description: "MAC (Media Access Control) address",
    name: "macaddr",
    alias: [],
    maxParams: 0,
  },
  {
    description: "currency amount",
    name: "money",
    alias: [],
    maxParams: 0,
  },
  {
    description: "exact numeric of selectable precision",
    name: "numeric",
    alias: ["decimal"],
    maxParams: 2,
  },
  {
    description: "geometric path on a plane",
    name: "path",
    alias: [],
    maxParams: 0,
  },
  {
    description: "PostgreSQL Log Sequence Number",
    name: "pg_lsn",
    alias: [],
    maxParams: 0,
  },
  {
    description: "geometric point on a plane",
    name: "point",
    alias: [],
    maxParams: 0,
  },
  {
    description: "closed geometric path on a plane",
    name: "polygon",
    alias: [],
    maxParams: 0,
  },
  {
    description: "single precision floating-point number (4 bytes)",
    name: "float4",
    alias: ["real"],
    maxParams: 0,
  },
  {
    description: "signed two-byte integer",
    name: "int2",
    alias: ["smallint"],
    maxParams: 0,
  },
  {
    description: "autoincrementing two-byte integer",
    name: "serial2",
    alias: ["smallserial"],
    maxParams: 0,
  },
  {
    description: "autoincrementing four-byte integer",
    name: "serial4",
    alias: [],
    maxParams: 0,
  },
  {
    description: "autoincrementing four-byte integer",
    name: "serial",
    alias: [],
    maxParams: 0,
  },
  {
    description: "variable-length character string",
    name: "text",
    alias: [],
    maxParams: 0,
  },
  {
    description: "time of day (no time zone)",
    name: "time",
    alias: ["time without time zone"],
    maxParams: 1,
  },
  {
    description: "time of day, including time zone",
    name: "timetz",
    alias: ["time with time zone"],
    maxParams: 0,
  },

  {
    description: "date and time, including time zone",
    name: "timestamptz",
    alias: ["timestamp with time zone"],
    maxParams: 1,
  },
  {
    description: "date and time (no time zone)",
    name: "timestamp",
    alias: ["timestamp without time zone"],
    maxParams: 1,
  },
  {
    description: "text search query",
    name: "tsquery",
    alias: [],
    maxParams: 0,
  },
  {
    description: "text search document",
    name: "tsvector",
    alias: [],
    maxParams: 0,
  },
  {
    description: "user-level transaction ID snapshot",
    name: "txid_snapshot",
    alias: [],
    maxParams: 0,
  },
  {
    description: "universally unique identifier",
    name: "uuid",
    alias: [],
    maxParams: 0,
  },
  {
    description: "XML data",
    name: "xml",
    alias: [],
    maxParams: 0,
  },

  // Object Identifier Types
  // https://www.postgresql.org/docs/8.1/datatype-oid.html
  {
    description: "numeric object identifier",
    name: "oid",
    alias: [],
    maxParams: 0,
  },
  {
    description: "function name",
    name: "regproc",
    alias: [],
    maxParams: 0,
  },
  {
    description: "function with argument types",
    name: "regprocedure",
    alias: [],
    maxParams: 0,
  },
  {
    description: "operator name",
    name: "regoper",
    alias: [],
    maxParams: 0,
  },
  {
    description: "operator with argument types",
    name: "regoperator",
    alias: [],
    maxParams: 0,
  },
  {
    description: "relation name",
    name: "regclass",
    alias: [],
    maxParams: 0,
  },
  {
    description: "data type name",
    name: "regtype",
    alias: [],
    maxParams: 0,
  },
  {
    description: "System Catalog",
    name: "pg_catalog",
    alias: [],
    maxParams: 0,
  },
] as const;

export type TypeCatalog = typeof types;
export type TypeOrAlias =
  | TypeCatalog[number]["name"]
  | TypeCatalog[number]["alias"][number];

export const typeOrAlias: TypeOrAlias[] = types
  .flatMap((t) => [...t.alias, t.name])

  // NOTE order is important for the parser, "bit varying" needs to come before
  // "bit", for it to accurately parse.
  .slice()
  .sort((a, b) => {
    if (a.length < b.length) {
      return 1;
    } else {
      return -1;
    }
  });

export const includesReferenceCatalog: TypeOrAlias[] = [
  "interval year",
  "interval month",
  "interval day",
  "interval hour",
  "interval minute",
  "interval second",
  "interval year to month",
  "interval day to hour",
  "interval day to month",
  "interval day to minute",
  "interval day to second",
  "interval hour to minute",
  "interval hour to second",
  "interval minute to second",
  "time with time zone",
  "timestamp with time zone",
  "timestamp",
  "smallint",
  "integer",
  "bigint",
  "boolean",
  "real",
  "bit varying",
  "double precision",
  "bit",
  "int",
  "interval",
  "decimal",
  "numeric",
  "varchar",
  "char",
  "character",
  "character varying",
];

export const defaultTypeMods = {
  bit: 1,
  "interval year": 4,
  "interval month": 2,
  "interval day": 8,
  "interval hour": 1024,
  "interval minute": 2048,
  "interval second": 4096,
  "interval year to month": 6,
  "interval day to month": 6,
  "interval day to hour": 1032,
  "interval day to minute": 3080,
  "interval day to second": 7176,
  "interval hour to minute": 3072,
  "interval hour to second": 7168,
  "interval minute to second": 6144,
  char: 1,
  character: 1,
};
