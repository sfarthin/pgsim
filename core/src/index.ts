import { Query } from "./toParser";
import toLinter, { LintError, LintOptions } from "./toLinter";
import toParser from "./toParser";
import toDDL from "./toDDL";
import { fromFiles, fromString } from "./toSql";
import toArray from "./iteratorToArray";
import toSchema, { toTableFields, Schema } from "./toSchema";
import { PGError, PGErrorCode } from "./errors";

export { PGErrorCode, toTableFields, LintOptions, toArray };

export function lintString(
  str: string,
  options: LintOptions | void,
  schema: Schema | void
): Iterator<LintError, void> {
  return toLinter(fromString(str), options, schema);
}

export function lintFiles(
  files: string[],
  options: LintOptions | void,
  schema: Schema | void
): Iterator<LintError, void> {
  return toLinter(fromFiles(files), options, schema);
}

export function lintStream(
  stream: Iterator<string>,
  options: LintOptions | void,
  schema: Schema | void
): Iterator<LintError, void> {
  return toLinter(stream, options, schema);
}

export function getSingleQuery(str: string): Query {
  const queries = toArray(toParser(fromString(str)));

  if (queries.length !== 1) {
    throw new PGError(PGErrorCode.INVALID, "Expected Single Query");
  }
  const { query } = queries[0];

  return query;
}

export function toSchemaFromString(str: string): Schema {
  return toSchema(toParser(fromString(str)));
}

export function toDDLFromString(str: string): string {
  return toDDL(toSchemaFromString(str));
}
