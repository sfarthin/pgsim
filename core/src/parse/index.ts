import { optional, mixed, array, string, guard, exact } from "decoders";
import pgParse from "./pgParse";
import { queryDecoder, Queries } from "./query";
import toOriginalText from "../toReader/toOriginalText";
import { SelectStmt } from "./selectStmt";
import { PGError, PGErrorCode } from "../errors";

export * from "./selectStmt";
export * from "./alterTableStmt";
export * from "./createStmt";
export * from "./constraint";
export * from "./query";
export * from "./constant";

export const parserResultDecoder = exact({
  // This is mixed because Error messages are hard to read if we do this here, we validate each query seperately
  query: array(mixed),
  stderr: optional(string),
  error: optional(mixed),
});

export const validateParsedResult = guard(parserResultDecoder);

export const validateQuery = guard(queryDecoder);

const getLocationFromUnSafeQuery = (q: any) => {
  // @ts-ignore - with the conditional chaining this is harmless, byt typescript is concerned
  return Object.values(q)?.[0]?.relation?.RangeVar?.location ?? null;
};

export default function parse(sql: string): Queries {
  const unsafeResult = pgParse(sql);
  const { query: queries, stderr, error } = validateParsedResult(unsafeResult);

  // Native error
  if (stderr || error) {
    throw error || new Error(stderr);
  }

  let index = 0;
  try {
    return queries.map((unSafeQuery: any, i) => {
      index = i;
      return validateQuery(unSafeQuery);
    });
  } catch (e) {
    try {
      const queryAsText = toOriginalText(sql, index);
      e.message = `Unable to understand query ${`\n\n${queryAsText}\n\n`} ${
        e.message
      }`;
    } catch (e) {
      e.message = `Unable to understand query ${index + 1} ${e.message}`;
    }
    throw e;
  }
}

export function parseSelect(str: string): SelectStmt {
  const queries = parse(str);

  if (queries.length !== 1 || !("SelectStmt" in queries[0])) {
    throw new PGError(PGErrorCode.NOT_EXISTS, "Expected a single SELECT");
  }

  return queries[0].SelectStmt;
}
