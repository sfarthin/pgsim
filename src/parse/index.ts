import { Stmt } from "../types";
import {
  transform,
  sequence,
  zeroToMany,
  whitespace,
  sqlStyleComment,
  cStyleComment,
  or,
  ResultType,
  Rule,
  Context,
  combineComments,
  endOfInput,
  FailResult,
  SuccessResult,
} from "./util";
import { getFriendlyErrorMessage } from "./error";
import { variableSetStmt } from "./variableSetStmt";
import { createEnumStmt } from "./createEnumStmt";
import { createSeqStmt } from "./createSeqStmt";
import { alterSeqStmt } from "./alterSeqStmt";
import { createStmt } from "./createStmt";
import { dropStmt } from "./dropStmt";
import { alterEnumStmt } from "./alterEnumStmt";
import { alterOwnerStmt } from "./alterOwnerStmt";
import { indexStmt } from "./indexStmt";
import { alterTableStmt } from "./alterTableStmt";
import { selectStmt } from "./selectStmt";
import { viewStmt } from "./viewStmt";
import { codeComments } from "./codeComments";
import { renameStmt } from "./renameStmt";
import { updateStmt } from "./updateStmt";

class ParseError extends Error {
  expectedStmt?: Stmt;
  // context: Context;
  // result: FailResult;
  constructor(context: Context, result: FailResult, expectedStmt?: Stmt) {
    super();
    this.message = getFriendlyErrorMessage(context, result);
    // this.context = context;
    // this.result = result;
    this.expectedStmt = expectedStmt;
  }

  toString() {
    return "error";
  }
}

const CommentStatement = transform(
  sequence([
    zeroToMany(whitespace),
    or([cStyleComment, sqlStyleComment]),
    zeroToMany(whitespace),
  ]),
  (v) => {
    return { value: { Comment: v[1] }, eos: null };
  }
);

export const stmt = transform(
  or([
    variableSetStmt,
    createEnumStmt,
    createSeqStmt,
    alterSeqStmt,
    createStmt,
    dropStmt,
    alterEnumStmt,
    alterOwnerStmt,
    indexStmt,
    alterTableStmt,
    selectStmt,
    viewStmt,
    renameStmt,
    updateStmt,

    // Standalone comments
    CommentStatement,
  ]),
  (v, ctx) => {
    if (!ctx.numStatements) {
      ctx.numStatements = 1;
    } else {
      ctx.numStatements++;
    }

    return v;
  }
);

export const stmts: Rule<Stmt[]> = transform(
  sequence([
    zeroToMany(stmt),
    endOfInput, // <-- This ensures we don't return before hitting the end of our SQL.
  ]),
  (v) => {
    const stmts = v[0];

    const output: Stmt[] = [];

    let previousStmtLocation = 0;
    for (let i = 0; i < stmts.length; i++) {
      const c = stmts[i];

      output.push({
        stmt: c.value,
        ...(c.eos?.firstSemicolonPos
          ? {
              stmt_len: c.eos.firstSemicolonPos - previousStmtLocation,
            }
          : {}),
        ...(previousStmtLocation
          ? { stmt_location: previousStmtLocation }
          : {}),
      });

      if (c.eos?.lastSemicolonPos != null) {
        previousStmtLocation = c.eos.lastSemicolonPos + 1;
      }
    }

    return output;
  }
);

function reduceComments(acc: Stmt[], stmt: Stmt): Stmt[] {
  const previousStmt = acc[acc.length - 1];

  // combine comment nodes and remove whitespace.
  if (
    previousStmt &&
    "Comment" in previousStmt.stmt &&
    "Comment" in stmt.stmt
  ) {
    return [
      ...acc.slice(0, -1),
      {
        stmt: {
          Comment: combineComments(
            previousStmt.stmt.Comment,
            stmt.stmt.Comment
          ),
        },
      },
    ];
    // Just lead whitespace
  } else if ("Comment" in stmt.stmt) {
    return [
      ...acc,
      {
        stmt: {
          Comment: stmt.stmt.Comment,
        },
      },
    ];
  } else {
    return acc.concat(stmt);
  }
}

/**
 * Used mostly of validation of parse method
 */
export function parseComments(inputSql: string) {
  const result = codeComments({
    str: inputSql,
    pos: 0,
    numStatements: 0,
  });

  if (result.type == ResultType.Success) {
    return result.value;
  }

  throw new Error("sql brokenz");
}

/**
 * Human-friendly version of the raw generated parser's parse() function, but
 * with much better error reporting, showing source line position where it
 * failed.
 */
export default function parse(
  context: Context,
  expectedAst: Stmt[] = []
): SuccessResult<Stmt[]> {
  const result = stmts(context);

  if (result.type == ResultType.Success) {
    result.value = result.value.reduce(reduceComments, []);
    return result;
  }

  const error = new ParseError(
    context,
    result,
    expectedAst[context.numStatements ?? 0]
  );

  throw error;
}
