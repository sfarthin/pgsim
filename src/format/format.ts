import { Stmt } from "~/types";
import parse from "../parse";
import { toSingleLineIfPossible } from "./util";
import { toString, PrintOptions, createFriendlyStmtError } from "./print";
import { SuccessResult } from "../parse/util";
import { assertError } from "../assertError";
import formatStmt from "./stmt";

/**
 * If we already have the parse result, we can
 * pass that in rather than re-parsing it.
 */
export default function format(
  input: SuccessResult<Stmt[]>,
  filename: string,
  opts: Omit<PrintOptions, "filename">
): string;
export default function format(
  input: string,
  filename: string,
  opts?: Omit<PrintOptions, "sql" | "filename">
): string;
export default function format(
  input: string | SuccessResult<Stmt[]>,
  filename = "unknown",
  opts?: Partial<PrintOptions>
): string {
  const result =
    typeof input === "string" ? parse(input, filename, opts) : input;
  const sql = typeof input === "string" ? input : opts?.sql ?? "";

  const codeBlock = result.value.flatMap((stmt, i) => {
    try {
      const formattedStmt = toSingleLineIfPossible(formatStmt(stmt));
      if (i === 0) {
        return formattedStmt;
      } else {
        // Lets add a newline between statements.
        return [[], ...formattedStmt];
      }
    } catch (e) {
      assertError(e);

      throw createFriendlyStmtError(stmt, e, {
        sql,
        filename,
        colors: true,
        ...opts,
      });
    }
  });

  return toString(codeBlock, {
    sql,
    filename,
    ...opts,
  });
}
