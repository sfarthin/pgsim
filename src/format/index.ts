import createStmt from "./createStmt";
import variableSetStmt from "./variableSetStmt";
import createEnumStmt from "./createEnumStmt";
import createSeqStmt from "./createSeqStmt";
import alterSeqStmt from "./alterSeqStmt";
import dropStmt from "./dropStmt";
import alterEnumStmt from "./alterEnumStmt";
import alterOwnerStmt from "./alterOwnerStmt";
import alterTableStmt from "./alterTableStmt";
import viewStmt from "./viewStmt";
import selectStmt from "./selectStmt";
import renameStmt from "./renameStmt";
import comment from "./comment";
import updateStmt from "./updateStmt";
import { Stmt, StatementType } from "../types";
import {
  toLineAndColumn,
  getSnippetWithLineNumbers,
  findNextToken,
} from "../parse/error";
import indexStmt from "./indexStmt";
import parse from "../parse";
import c from "ansi-colors";
import { NEWLINE, textFormatter, Formatter } from "./util";

type Opts = {
  ignore?: StatementType[];
  ignoreAllExcept?: StatementType[];
  sql?: string; // <-- Original sql before formatting. useful when there is an error.
  filename?: string;
};

// const schemaStatements: StatementType[] = [
//   "CreateStmt",
//   "AlterTableStmt",
//   "CreateSeqStmt",
//   "VariableSetStmt",
//   "CreateEnumStmt",
//   "AlterSeqStmt",
//   "IndexStmt",
//   "CommentStmt",
// ];

function formatStmt<T>(stmt: Stmt, f: Formatter<T>, opts?: Opts): T[][] {
  const s = stmt.stmt;
  const statementType = Object.keys(s)[0] as StatementType;

  if (opts?.ignore && opts.ignore.includes(statementType)) {
    return comment(`-- Ignoring ${statementType} statements`, f);
  }

  if (opts?.ignoreAllExcept && !opts.ignoreAllExcept.includes(statementType)) {
    return comment(`-- Ignoring ${statementType} statements`, f);
  }

  if ("CreateStmt" in s) {
    return createStmt(s.CreateStmt, f);
  } else if ("DropStmt" in s) {
    return dropStmt(s.DropStmt, f);
  } else if ("VariableSetStmt" in s) {
    return variableSetStmt(s.VariableSetStmt, f);
  } else if ("CreateEnumStmt" in s) {
    return createEnumStmt(s.CreateEnumStmt, f);
  } else if ("CreateSeqStmt" in s) {
    return createSeqStmt(s.CreateSeqStmt, f);
  } else if ("AlterSeqStmt" in s) {
    return alterSeqStmt(s.AlterSeqStmt, f);
  } else if ("AlterEnumStmt" in s) {
    return alterEnumStmt(s.AlterEnumStmt, f);
  } else if ("AlterOwnerStmt" in s) {
    return alterOwnerStmt(s.AlterOwnerStmt, f);
  } else if ("IndexStmt" in s) {
    return indexStmt(s.IndexStmt, f);
  } else if ("AlterTableStmt" in s) {
    return alterTableStmt(s.AlterTableStmt, f);
  } else if ("Comment" in s) {
    return comment(s.Comment, f);
  } else if ("SelectStmt" in s) {
    return selectStmt(s.SelectStmt, f);
  } else if ("ViewStmt" in s) {
    return viewStmt(s.ViewStmt, f);
  } else if ("RenameStmt" in s) {
    return renameStmt(s.RenameStmt, f);
  } else if ("UpdateStmt" in s) {
    return updateStmt(s.UpdateStmt, f);
  }

  throw new Error(`Cannot format ${Object.keys(s)}`);
}

export default function format(_stmts: Stmt[] | string, opts?: Opts): string {
  const stmts = typeof _stmts === "string" ? parse(_stmts) : _stmts;
  const f = textFormatter;

  if (stmts.length === 0) {
    return "";
  }

  return stmts
    .map((stmt) => {
      try {
        return f.concat(formatStmt(stmt, f, opts));
      } catch (_e) {
        const e = _e as { name?: string; message?: string };
        const errorType = e.name === "Error" ? "" : `(${e.name})`;
        if (opts?.filename && opts?.sql) {
          // Lets skip over any comments
          const { start: pos } = findNextToken(
            opts.sql,
            stmt.stmt_location ?? 0
          );

          const { line, column } = toLineAndColumn(opts.sql, pos);

          e.name = `Problem formatting${errorType} ${c.cyan(
            opts.filename
          )}(${c.cyan(String(line + 1))},${c.cyan(String(column + 1))})`;

          e.message = `${
            e.message
          }${NEWLINE}${NEWLINE}${getSnippetWithLineNumbers({
            str: opts.sql,
            start: pos,
            end: (stmt.stmt_location ?? 0) + (stmt.stmt_len ?? 20),
          })}${NEWLINE}`;
        } else {
          e.name = `Problem formatting${errorType}`;
          e.message = `${e.message}${NEWLINE}${NEWLINE}${JSON.stringify(
            stmt,
            null,
            2
          )}`;
        }
        throw e;
      }
    })
    .join(`${NEWLINE}${NEWLINE}`);
}
