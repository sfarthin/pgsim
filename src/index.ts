import parse from "~/nativeParse";
import toLinter, { LintError, LintOptions } from "~/toLinter";
import {
  fromFilesToSqlIterator,
  fromStringToSqlIterator,
} from "~/util/toSqlIterator";
import toArray from "~/util/iteratorToArray";
import toSchema, { toTableFields, Schema } from "~/toSchema";
import { PGErrorCode } from "~/util/errors";

export { PGErrorCode, toTableFields, LintOptions, toArray };

export function lintString(
  str: string,
  options: LintOptions | void,
  schema: Schema | void
): Iterator<LintError, void> {
  return toLinter(fromStringToSqlIterator(str), options, schema);
}

export function lintFiles(
  files: string[],
  options: LintOptions | void,
  schema: Schema | void
): Iterator<LintError, void> {
  return toLinter(fromFilesToSqlIterator(files), options, schema);
}

export function lintStream(
  stream: Iterator<string>,
  options: LintOptions | void,
  schema: Schema | void
): Iterator<LintError, void> {
  return toLinter(stream, options, schema);
}

export function toSchemaFromString(str: string): Schema {
  return toSchema(parse(str));
}

export { parse };
