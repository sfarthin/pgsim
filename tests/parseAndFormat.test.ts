import parse from "../src/parse";
import nParse from "../src/nativeParse";
import format from "../src/format";
import omitDeep from "./omitDeep";
import { Stmt } from "../src/types";
import { readFileSync, readdirSync } from "fs";
import { join } from "path";

/**
 * By ensuring the native parser is verified by the decoder
 * and the PEG.js parser matches the base output of the
 * native parser, we can verify the PEG.js is a valid superset
 */

const files: { [s: string]: string } = readdirSync(
  join(__dirname, "./parseAndFormat")
).reduce((acc, file) => {
  if (!file.match(/\.sql$/)) {
    return acc;
  }
  return {
    ...acc,
    [file]: readFileSync(join(__dirname, "./parseAndFormat", file)).toString(),
  };
}, {});

function removeComments(stmts: Stmt[]): Stmt[] {
  return (
    (stmts.map((stmt) =>
      omitDeep(stmt as object, [
        // Only in new parser
        "comment",
      ])
    ) as Stmt[])
      // Only in PEGJS parser
      .filter((stmt) => !("Comment" in stmt.RawStmt.stmt))
  );
}

function removeStyle(stmts: Stmt[]): Stmt[] {
  return stmts.map((stmt) =>
    omitDeep(stmt as object, [
      // These vary slightly from the native parser and our parser
      "stmt_len",
      "stmt_location",

      // different in each parser
      "location",
    ])
  ) as Stmt[];
}

function checkParserAndFormatter(sql: string, filename: string): void {
  const ast = parse(sql, filename);
  const realAst = nParse(sql);

  // Compare two ASTs
  // console.log(JSON.stringify(realAst, null, 2), JSON.stringify(ast, null, 2));

  // console.log(format(ast));
  // console.log(sql);

  // Make sure parser is identical to native parser
  expect(removeComments(ast)).toEqual(removeComments(realAst));

  // Make sure formatting and parsing produces the same AST
  expect(removeStyle(parse(format(ast)))).toEqual(removeStyle(ast));

  // Lets make sure we are aware of any changes
  expect(format(ast)).toMatchSnapshot();
}

describe("Parse and format", () => {
  for (const file in files) {
    // if (file === "createStmt.sql") {
    it(file, () => checkParserAndFormatter(files[file], file));
    // }
  }
});
