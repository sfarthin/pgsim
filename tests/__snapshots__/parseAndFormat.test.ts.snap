// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse and format alterSeqStmt.sql 1`] = `
"-- Hello
ALTER SEQUENCE foo 
	OWNED BY a.aaa;

-- Hello
ALTER SEQUENCE foo 
	-- foo
	-- foo2
	INCREMENT BY 1;
"
`;

exports[`Parse and format createEnumStmt.sql 1`] = `
"-- Top 1
-- Top 2

-- 1 
--  2
-- 3
-- 4
-- 5
--  6 
-- 7
-- 8
-- 9
--  10 
-- 11
CREATE TYPE foo AS ENUM (
	-- comment on a
	'a',
	-- comment on b
	'b',
	-- comment on c
	'c',
	-- comment on d
	'd',
	-- comment on e
	'e'
);

-- Bottom 1
-- Bottom 2
"
`;

exports[`Parse and format createSeqStmt.sql 1`] = `
"-- This has no options
CREATE SEQUENCE foo;

-- 1 
-- 2
CREATE SEQUENCE foo 
	-- Start With 1
	START WITH 1
	-- INCREMENT BY 1
	INCREMENT BY 1
	NO MINVALUE
	NO MAXVALUE
	NO CYCLE
	CACHE 1
	-- foo bac
	OWNED BY NONE;

-- another comment
-- and another
-- foo bar

CREATE SEQUENCE IF NOT EXISTS foo2 
	MINVALUE 4
	MAXVALUE 6
	-- different variation
	INCREMENT BY 2
	-- different variation
	START WITH 2
	CYCLE
	OWNED BY foo.bar;
"
`;

exports[`Parse and format createStmt.sql 1`] = `
"-- CREATE /*foo */ TABLE adsiiodnsaoidnoasindosaind9 -- foo
-- (
--     -- integer
--     ahahhahah integer,
--     ahahhahah integer, /* asdasd */ /* asasd */ -- asd
--     aa integer NOT NULL DEFAULT 1 PRIMARY KEY,
--     aab int DEFAULT '4', -- int
--     /* int4 */
--     aac int4,
--     aa bigint DEFAULT 1,
--     /* 1 */ ab /* 2 */ int8 /* 3 */ NOT /* 4 */ NULL /* 5 */, -- 6
--     /* 1 */ ac /* 2 */ bigserial /* 3 */ NULL /* 4 */, -- 5
--     ad serial8 UNIQUE,
--     ae serial,
--     af serial4,
--     aba bit,
--     abb bit(2),
--     abc bit varying,
--     abca bit varying(3),
--     abd varbit,
--     abda varbit(3),
--     bba interval,
--     bbb interval day to hour,
--     ba bool DEFAULT true,
--     bbaa boolean,
--     caa box,
--     cba bytea,
--     cc /* hello */ cidr,
--     cd /* yo */ circle /* yo */,
--     ce date,
--     cf inet,
--     cg line,
--     ch lseg,
--     ci macaddr,
--     cj money,
--     ck tsquery,
--     cl tsvector,
--     cm txid_snapshot,
--     cn uuid,
--     co xml,
--     daa real,
--     dba float4,
--     dda smallint,
--     ddb int2,
--     da double precision,
--     db float8,
--     eaa decimal,
--     eb numeric,
--     eca decimal(5),
--     ecb numeric(5),
--     eccv decimal(5,2),
--     ecd numeric(5,2),
--     bb character varying(255),
--     bc character varying(4),
--     bd character varying,
--     be varchar,
--     bf varchar(235),
--     bg character,
--     bh character(4),
--     bi char,
--     bj char(4),
--     bk text,
--     caz timestamp with time zone,
--     cbx timestamptz,
--     cax time with time zone,
--     cbxx timetz
-- );

CREATE TABLE some_table (
	-- This comment will go with \\"c'
	c INTEGER
);
-- create table reference_examples (
--     a integer references products,
--     /* 1 */ bss /* 2 */ integer /* 3 */ references /* 4 */ products /* 5 */ ( /* 6 */some_id /* 7 */), -- 8
--     c integer,
--     FOREIGN KEY (c) REFERENCES other_table (c1)
-- )
-- create table reference_group_examples (
--     a integer references products,
--     b integer,
--     c integer,
--     FOREIGN KEY (b,c) REFERENCES other_table (c1, c2)
-- )
"
`;

exports[`Parse and format variableSetStmt.sql 1`] = `
"SET statement_timeout = 0;

SET lock_timeout = 0;

SET client_encoding = 'UTF8';

SET standard_conforming_strings = 'on';

SET check_function_bodies = 'false';

SET client_min_messages = 'warning';

-- 1 
-- foo
SET statement_timeout = 0;

SET client_encoding = ' UTF8';
"
`;
